File: src/compiler.cc

#include "compiler.h"
#include "utils.h"
#include "templates.h"

using namespace std;

const string COMPILATION_TEXT = "Compilation";

bool compile_file(const Compiler& compiler, const filesystem::path& source, const filesystem::path& output, const filesystem::path& errors) {
    DEBUG("Compiling " + source.string() + " to " + output.string());
    if (not filesystem::exists(source)) {
        DEBUG(source.string() + " doesn't exist!");
        return false;
    }
    
    if (filesystem::exists(output)) {
        DEBUG("Removing previous compilation output: " + output.string());
        filesystem::remove(output);
    }

    if (filesystem::exists(errors)) {
        DEBUG("Removing previous compilation error output: " + errors.string());
        filesystem::remove(errors);
    }

    string command = compiler.command + " " + compiler.flags + " -c " + source.string() + " -o " + output.string() + " 2> " + errors.string();
    
    return run_system_command(command) == 0 and filesystem::exists(output);
}

bool link_file(const Compiler& compiler, const filesystem::path& source, const filesystem::path& output, const filesystem::path& errors) {
    DEBUG("Linking " + source.string() + " to " + output.string());
    if (not filesystem::exists(source)) {
        DEBUG(source.string() + " doesn't exist!");
        return false;
    }

    if (filesystem::exists(output)) {
        DEBUG("Removing previous linking output: " + output.string());
        filesystem::remove(output);
    }

    if (filesystem::exists(errors)) {
        DEBUG("Removing previous linking error output: " + errors.string());
        filesystem::remove(errors);
    }

    string command = compiler.command + " " + compiler.flags + " " + source.string() + " -o " + output.string() + " 2> " + errors.string();;
    
    return run_system_command(command) == 0 and filesystem::exists(output);
}

bool check_p1xx_compiles(const Problem& p, filesystem::path& errors) {
    DEBUG("Checking if P1++ compiles the user's main.cc...");
    filesystem::path object_file = p.advocat_dir / "main.o";
    errors = p.advocat_dir / "p1xx-compilation.err";

    return compile_file(P1XX, p.source, object_file, errors);
}

bool compile_binary(const Problem& p, const filesystem::path& templates_dir, filesystem::path& errors) {
    DEBUG("Compiling the binary for testing with G++...");
    filesystem::path joined_source = p.advocat_dir / "joined.cc";

    string template_name;
    if (p.has_main) template_name = "normal.cc.in";
    else template_name = "nomain.cc.in";
    apply_template(template_name, templates_dir, joined_source, p);

    filesystem::path object_file = p.advocat_dir / "joined.o";

    errors = p.advocat_dir / "c1xx-compilation.err";
    bool compiles = compile_file(GXX, joined_source, object_file, errors);
    if (not compiles) return false;

    errors = p.advocat_dir / "c1xx-linking.err";
    return link_file(GXX, object_file, p.output, errors);
}

bool compile_problem(const Problem& p, const filesystem::path& templates_dir) {
    show_task_status(COMPILATION_TEXT, TaskType::Test, TaskStatus::InProgress);

    filesystem::path errors;
    bool compiles = check_p1xx_compiles(p, errors) and compile_binary(p, templates_dir, errors);

    show_task_status(COMPILATION_TEXT, TaskType::Test, compiles ? TaskStatus::Pass : TaskStatus::Fail);

    if (not compiles) {
        string details;
        read_file(errors, details);
        show_details("Compilation output", details);
    }

    return compiles;
}

File: src/compiler.rs
use std::{fmt, io, path, process};
use crate::{debug, problem};

pub static P1XX: Compiler = Compiler {
    command: "g++",
    flags1: &["-D_JUDGE_", "-DNDEBUG", "-O2", "-Wall", "-Wextra", "-Werror", "-Wno-sign-compare", "-Wshadow"],
    flags2: &["-D_JUDGE_", "-DNDEBUG", "-O2"]
};

pub struct Compiler<'a> {
    command: &'a str,
    flags1: &'a[&'a str],
    flags2: &'a[&'a str]
}

pub enum CompilationError {
    SourceDoesNotExist,
    SourceIsADir,
    OutputIsADir,
    ExecutionError(io::Error),
    CompilerError(String),
    MissingOutput
}

impl fmt::Display for CompilationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            CompilationError::SourceDoesNotExist => write!(f, "Source file doesn't exist!"),
            CompilationError::SourceIsADir => write!(f, "Source is a directory!"),
            CompilationError::OutputIsADir => write!(f, "Output is a directory!"),
            CompilationError::ExecutionError(e) => write!(f, "Command execution error: {}", e),
            CompilationError::CompilerError(stderr) => write!(f, "The compiler raised an error:\n{}", stderr),
            CompilationError::MissingOutput => write!(f, "Can't find the compiler output!")
        }
    }
}

pub struct CompileProcessError {
    pub pass: u8,
    pub error: CompilationError
}

impl fmt::Display for CompileProcessError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Pass {}: {}", self.pass, self.error)
    }
}

enum CompilationType {
    Object,
    Binary
}

impl Compiler<'_> {
    fn run(&self, source: &path::Path, output: &path::Path, compilation_type: CompilationType, flags: &[&str]) -> Result<(), CompilationError>{
        if !source.exists() {
            Err(CompilationError::SourceDoesNotExist)
        } else if source.is_dir() {
            Err(CompilationError::SourceIsADir)
        } else if output.is_dir() {
            Err(CompilationError::OutputIsADir)
        } else {
            let mut command = process::Command::new(&self.command);
            command.args(flags)
                .args(["-o", output.to_string_lossy().as_ref()]);

            match compilation_type {
                CompilationType::Object => command.args(["-c", source.to_string_lossy().as_ref()]),
                CompilationType::Binary => command.arg(source.to_string_lossy().to_string())
            };

            debug!("Running command: {:?}", command);
            let command = command.output().map_err(CompilationError::ExecutionError)?;

            if command.status.success() {
                if output.is_file() {
                    Ok(())
                } else {
                    Err(CompilationError::MissingOutput)
                }
            } else {
                Err(CompilationError::CompilerError(String::from_utf8_lossy(&command.stderr).to_string()))
            }
        }
    }

    fn compile_first_pass(&self, source: &path::Path, output: &path::Path) -> Result<(), CompilationError> {
        self.run(source, output, CompilationType::Object, self.flags1)
    }

    fn compile_and_link_first_pass(&self, source: &path::Path, output: &path::Path) -> Result<(), CompilationError> {
        self.run(source, output, CompilationType::Binary, self.flags1)
    }

    fn compile_and_link_second_pass(&self, source: &path::Path, output: &path::Path) -> Result<(), CompilationError> {
        self.run(source, output, CompilationType::Binary, self.flags2)
    }

    pub fn compile_problem(&self, problem: &problem::Problem, generated_source: &path::Path) -> Result<(), CompileProcessError>{
        debug!("Running the first pass compilation (P1++ checks)");
        if problem.has_main {
            let output = problem.tmp_dir.join("main.x");
            self.compile_and_link_first_pass(problem.source.as_path(), output.as_path())
        } else {
            let output = problem.tmp_dir.join("main.o");
            self.compile_first_pass(problem.source.as_path(), output.as_path())
        }.map_err(|error| CompileProcessError {pass: 1, error})?;

        debug!("Running the second pass compilation (G++ binary)");
        self.compile_and_link_second_pass(generated_source, problem.output.as_path())
            .map_err(|error| CompileProcessError {pass: 2, error})
    }
}